### DISCUSSION TITLE
Localization/translation

### DISCUSSION THREAD

**Discussion by terjeio** [terjeio/ioSender#140] [General]
Initial support for localization/translation support has been added to the latest version. The default locale is en-US (US english), and a partial translation to zh-CN  (simplified chinese) has been added by me.

I will not make any translations myself but rather rely on the community to provide them. See the [Wiki](https://github.com/terjeio/ioSender/wiki/Localization) for how this can be done.

I suggest that those who want to translate add to this discussion before starting so that duplicate work can be avoided.

Tips: If you want to run ioSender with a specific language then create a shortcut to the executable and add the locale specification to the [command line](https://github.com/terjeio/ioSender/wiki/Setup-and-configuration). E.g. `-locale en-US` will select the default locale independent of what the OS default is.

**Comment by einencool** [comment]
Sounds Great. 
As discussed last year, I would like to help with the German Translation

**Reply by einencool** [reply]
Is it allowed to use signs like "Ü" "Ä" and so on? 
These signs are used very often in the german language?!

**Reply by terjeio** [reply]
> Is it allowed to use signs like "Ü" "Ä" and so on?

Yes, this since the [UTF-8 character encoding](https://en.wikipedia.org/wiki/UTF-8) is used.

**Reply by einencool** [reply]
This is the First test for the de-DE Version, if I understood it correct, then I need to upload the "translation.txt" so that you can make the "new dll files"...
[translation.txt](https://github.com/terjeio/ioSender/files/7444012/translation.txt)

**Reply by terjeio** [reply]
Here are the dll files, unpack to the sender folder:

[de-DE.zip](https://github.com/terjeio/ioSender/files/7444370/de-DE.zip)

As expected some strings are too long - some I can reserve more space for quite easily, some perhaps not. I'll look into that later...

**Reply by einencool** [reply]
I‘ll have a look at it tomorrow. Is there a „dummy machine configuration“ so that I can test ALL features?
At Home I don‘t have my machine, and here I‘ve only the old Teensy 4.0 Board.
Thank you for your quick reply, then I can make some tests together with my father, because he can‘t speak english or anything else :-)

**Reply by einencool** [reply]
Here are some Pictures, where the translation doesn't work or the size is to big. (it's only a short test with the teensy 4.0 Board and an old GRBLHal Version.
[Bilder.zip](https://github.com/terjeio/ioSender/files/7446583/Bilder.zip)

Short Discription for the color in the pictures:
Yellow - Please adjust the width of the buttons
Red - I'll change the translation maybe back to the original english words or something shorter
Cyan / green - These areas are not translated 

In the Probing Area there are many fields not translated, also on the right side the "Fly Out Bar" is not translated
And All settings from GRBL are without ANY information, but this is also in the english version, so maybe this is because of the old GRBLHal Version

**Reply by terjeio** [reply]
> In the Probing Area there are many fields not translated, also on the right side the "Fly Out Bar" is not translated

I'll go through all this, some strings I know are missing since they are not part of the UI definition - there is a workaround for those I have to add.

> And All settings from GRBL are without ANY information, but this is also in the english version, so maybe this is because of the old GRBLHal Version

Yes - in the later version everything in this tab is fetched from the controller. Any translations of settings has to be made in a file that the sender will load. I have basic functionality in place for that - a workflow of how to do it will have to be put in place. In the latest versions of the controller I have added $-commands for extracting the strings to files. Error and alarm codes/strings are also part of this. The [config app](https://github.com/terjeio/ioSender/releases/download/2.0.34/Config.App.2.0.34.zip) has file-menu entries to simpify extracting/saving the data.

**Reply by terjeio** [reply]
@einencool I have uploaded [ioSender Edge.36p3](http://www.io-engineering.com/downloads) with some of the adjustments made. The _locale_ subfolder contains updated files with some of the missing strings added.

I have not yet added translations for the lathe related tabs, to be added later when the main part is done. I assume not many are using these anyway...

It will take me some time to go through all non-UI defined strings (not in xaml files) and make them available for translation, so be patient. Translation is not something done quickly.

Note that from this edge version the gcode-parser has been changed significantly. It is now possible to reference both numbered and named parameters and use these in expressions. I have not yet tested parameter and expression support extensively, there could be bugs lurking. "Normal" gcode should work as before as the parsed version is mainly used for 3D rendering. A teaser:
```
G90G21
#1=3
#<test>=[[#1+SIN[45]]*10]
G0X0Y0
G1X[#<test>]F200
G1Y[#<test>]
G1X0
G1Y0
M30
```

**Reply by einencool** [reply]
Can you explain what the "Dent length" for the drag knife means? Dent has many tranlations, and I don't know what it could be
I'm starting to make some new translations, and hope it could fit.
Hopefully I get it in the next hour, because then I have to leave :-)

Maybe I add some new questions for meanings, but then I'll edit this post.

**Reply by terjeio** [reply]
See line 33+ in the attached file for explanations, I've used this as the basis for porting to ioSender.

[Drag-Knife Script.txt](https://github.com/terjeio/ioSender/files/7448489/Drag-Knife.Script.txt)

**Reply by einencool** [reply]
So, this is what I could get till now, there are some fields missing like you mentioned before, but it seems to work very good :-)
Thank you for your great work :-)
Now I have to leave, maybe in the evening I can make some more tests otherwise we will see next week ;-)

[translation.txt](https://github.com/terjeio/ioSender/files/7448587/translation.txt)

Here are some Screenshots what I found which is not in the translation for now:
![image](https://user-images.githubusercontent.com/37541729/139579117-c715712b-dbca-44d6-8c96-6e4a52a8066a.png)

![image](https://user-images.githubusercontent.com/37541729/139579177-c772611a-ec93-485a-83d4-9385b3395cff.png)

![image](https://user-images.githubusercontent.com/37541729/139579345-af4711fa-d263-403f-86e4-fe89711a07b0.png)

When in Settings:GRBL Tab, this Prompt comes up:
![image](https://user-images.githubusercontent.com/37541729/139579385-c62a97e4-9340-450c-905f-7ae980edb9d9.png)

Tooltips in these areas are not translated:
![image](https://user-images.githubusercontent.com/37541729/139579247-e85ede12-07ae-40a9-9143-fd22476f06da.png)

**Reply by terjeio** [reply]
Thanks, I have uploaded a new version with your changes and with some new strings to translate.

The missing tooltips above are due to prorgam logic changing them at run time - to be fixed later.
The _Status:_ `IDLE` has the `IDLE` text (and the other statuses) coming from the controller as plaintext. I'll have to figure out how to translate those, perhaps via a translatable lookup table...

From here it will take a little longer before I will be able to deliver any updates.

**Reply by einencool** [reply]
Alright, I think you've done an amazing job 👍, for the "normal" operation it seems to work really good, and there are only a few things to do...
I did some minor Fixes and the few new translations.
Your new 36.4 Version was still named as 36.3 ;-), but that's just cosmetic ;-)
[translation 36.4.txt](https://github.com/terjeio/ioSender/files/7449230/translation.36.4.txt)

Here are the Screenshots, so that you can see the points that have to be a little bit improved :-)
Outline Menu:
![image](https://user-images.githubusercontent.com/37541729/139591603-6bf9460f-85f2-4aea-b59f-28b6dc21552f.png)

3D View from Mainwindow is correct,  But from the 3D-View TAB there is this Box not in the correct size:
![image](https://user-images.githubusercontent.com/37541729/139591689-7ff93c15-72cd-49ad-a398-8eead4157863.png) ![image](https://user-images.githubusercontent.com/37541729/139591727-125e0c54-583a-4ed7-8c36-6336bcbd5b16.png)

Tool Tips for Signals in Main Window:
![image](https://user-images.githubusercontent.com/37541729/139591774-25559de3-074f-40af-8fdb-d549e8f6956a.png)

![image](https://user-images.githubusercontent.com/37541729/139592629-3bae78d7-8dc0-42b1-a2a7-b0bcc4d99cd3.png)

Size for translation in Center Finder:
![image](https://user-images.githubusercontent.com/37541729/139592233-46f87290-3e01-4ed4-830a-4effe5468d6d.png)

Area to Probe in height map:
![image](https://user-images.githubusercontent.com/37541729/139592326-3b921760-129e-4e7b-ae32-0742d93fb9ee.png)

different Hints in Probing tabs:
![image](https://user-images.githubusercontent.com/37541729/139592478-c7f11212-919b-4309-9099-ea2877045158.png)

![image](https://user-images.githubusercontent.com/37541729/139592489-e726105e-627a-47bd-81ae-f7d15747a03f.png)

![image](https://user-images.githubusercontent.com/37541729/139592505-7800d385-efe6-4449-934d-e26926043728.png)

![image](https://user-images.githubusercontent.com/37541729/139592521-698f87a5-3a4b-4a73-99d5-e3a4506bcf15.png)

![image](https://user-images.githubusercontent.com/37541729/139592585-4de2ab0c-1aca-4f64-9739-0c6b655f000c.png)

Complete GRBL Settings Tab:
![image](https://user-images.githubusercontent.com/37541729/139592363-adc0b955-b149-4d45-89df-f0151a315f19.png)

Settings Program Tab:
![image](https://user-images.githubusercontent.com/37541729/139592438-9bdefc0e-9aec-4e6c-bb20-dbf6960456c3.png)

Thank you, greets Chris

Oh, one thing, is there a way to start the Software without a connected  Board, so that all tabs are visible and clickable?
Because without the Board everything is grey and sadly not clickable :-(
![image](https://user-images.githubusercontent.com/37541729/139593084-132730fa-4219-4c98-bc1f-64b3d932ed48.png)

**Reply by terjeio** [reply]
New version uploaded. I have cracked moving strings from code to UI definition files - 2 strings moved this far, a lot more to go...
Until I get this done you do not need to mark any screen dumps for me.

> Oh, one thing, is there a way to start the Software without a connected Board, so that all tabs are visible and clickable?

Currently not, and I am not sure it ever will be possible due to the tight integration between the controller and the sender I want to have. The closest will be to have a "soft" controller for the basic parts. The [simulator driver](https://github.com/grblHAL/Simulator) fits the bill but I have only managed to compile it for Linux for now.

**Reply by phil-barrett** [reply]
> > Oh, one thing, is there a way to start the Software without a connected Board, so that all tabs are visible and clickable?


> 
> Currently not, and I am not sure it ever will be possible due to the tight integration between the controller and the sender I want to have. The closest will be to have a "soft" controller for the basic parts. The simulator driver fits the bill but I have only managed to compile it for Linux for now.

Kind of a kludge but you can put grbl on a cheap arduino and hook it up.  You could also put grblHAL on a teensy (or any other supported micro) and use that - not quite as cheap. You don't need a full cnc machine.

**Reply by terjeio** [reply]
...or a cheap(er) Pi Pico? We could provida a binary for that as it can be programmed by drag & drop, no need to install anything.

**Comment by terjeio** [comment]
[New upload](http://www.io-engineering.com/downloads) with what I believe is a near 100% extract of translatable strings (exept Lathe tabs/settings - I'll add these later).

Note that many of the new strings has placeholders that is used for inserting run-time data. These are `{0}` ,`{1}` etc and must be kept in the translated strings.

Some special characters are encoded as [html character entities](https://www.w3schools.com/html/html_entities.asp), eg `&lt;` and `&gt;`, commmon for these is that they starts with an ampersand and ends with a semicolon. Again these should be kept.

Finally, some strings contain line breaks, these are represented by `'\n'`.

Initial translation to russian has been provided by @calabr in issue #145, this is incorporated in the new upload.

**Reply by einencool** [reply]
Here's the new Version, I made some minor bugfixes.
It seems, that these Tooltips are not in the translation List:
![image](https://user-images.githubusercontent.com/37541729/140536517-8a4440b1-f8d7-435a-9fd3-e57e2b61dee2.png)

The rest seems to be fine, and I hope my translations make sense in the most points :-:
[translation 36.7.txt](https://github.com/terjeio/ioSender/files/7487172/translation.36.7.txt)

**Reply by terjeio** [reply]
Sorry about those, I have added them now.
I have also added support for parameters and expressions in macros - hopefully that does not break any existing macros.

**Comment by calabr** [comment]
It's will be good to know the maximum character count available, especially for buttons.
 Direct Russian translation I made in most cases longer than English.
Also some labels need to enable line wrapping
![image](https://user-images.githubusercontent.com/64111035/140613499-d65afcd1-46ad-4913-bfac-05eff9f671f6.png)

**Reply by terjeio** [reply]
> It's will be good to know the maximum character count available, especially for buttons.

There is no character limit, widths are defined in pixels - and the size of any given string is not possible to know without measuring it with the correct font settings. I have made changes to better accommodate some of them (dynamic width adjustments within limits).

> Also some labels need to enable line wrapping

Wrapping is not available for labels, I _can_ add a line break in the original source - but when translated those are lost.
I had to change the label you have marked above to at TextBlock to allow wrapping.

**Comment by terjeio** [comment]
[New version](http://www.io-engineering.com/downloads) has been uploaded.

Translation of alarms, error messages and settings is now possible - I will add details about how in a Wiki, likely the one for the grblHAL core.

**Reply by einencool** [reply]
Sounds good, here are the translations for 36.8
[translation 36.8.txt](https://github.com/terjeio/ioSender/files/7492216/translation.36.8.txt)
